### PR TITLE
feat: add CI to detect references

### PR DESCRIPTION
**Pull request by SMillerDev**
This should tell us when references are no longer in use.

### CODE REVIEW

**Comment by issyl0** [comment]
Looks like some of them have `.patch` or `.diff` extensions:

```
Formula/w/webfs.rb:    url "https://github.com/Homebrew/formula-patches/raw/0518a6d1ed821aebf0de4de78e39b57d6e60e296/webfs/patch-ls.c"
Formula/y/yconalyzer.rb:    url "https://raw.githubusercontent.com/Homebrew/formula-patches/85fa66a9/yconalyzer/1.0.4.patch"
Formula/z/zyre.rb:    url "https://raw.githubusercontent.com/Homebrew/formula-patches/03cf8088210822aa2c1ab544ed58ea04c897d9c4/libtool/configure-big_sur.diff"
```

File: .github/workflows/check-reference.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+      - name: Detect references
+        run: |
+          patches=$(find -name "*.patch" -printf "%f\n")
+          status=0
+          for patch in $patches; do
+            if ! git -C "homebrew-core" grep -r "${patch/./}" > /dev/null; then
+              echo "$patch not needed!"
+              status=1
+            fi
+          done
```
Comment (issyl0) [inline .github/workflows/check-reference.yml:33]: Maybe:

```shell
#! /bin/bash

core_patches=$(git -C "$(brew --repo homebrew/core)" grep -h "Homebrew/formula-patches" | awk -F/ '{print $NF}' | tr -d '"')
formula_patches_here=$(git ls-files | grep -v "LICENSE" | grep -v "README.md" | grep -v ".github/workflows/triage-issues.yml")
status=0
for patch in $formula_patches_here; do
  if ! grep -q "$patch" <<< "$core_patches"; then
    echo "Unused patch: $patch."
    status=1
  fi
done
exit $status
```

Comment (issyl0) [reply .github/workflows/check-reference.yml:33]:
```shell
    echo "Unused patch: $patch; deleting."
    rm "$patch"
```

And then the GitHub Actions add and push changes stuff we have in other workflows?

Comment (issyl0) [reply .github/workflows/check-reference.yml:33]: Wait, no, I got these the wrong way around. 😩

Comment (issyl0) [reply .github/workflows/check-reference.yml:33]: It's actually:

```shell
#!/bin/bash

core_patches=$(git -C "$(brew --repo homebrew/core)" grep -h "Homebrew/formula-patches" | awk -F/ '{print $NF}' | tr -d '"')
all_patches=$(git ls-files | grep -v "LICENSE" | grep -v "README.md" | grep -v ".github/workflows/triage-issues.yml")
status=0

for patch in $all_patches; do
# If patch is not in core patches, it is unused.
  if ! grep -q "$core_patches" <<< "$patch"; then
    echo "Unused patch: $patch."
    status=1
  fi
done
exit $status
```

Comment (issyl0) [reply .github/workflows/check-reference.yml]: I pushed a commit - hope you don't mind!

File: .github/workflows/check-reference.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        with:
+          repository: 'homebrew/homebrew-core'
+          path: homebrew-core
+
+      - name: Detect references
+        run: |
+          #!/bin/bash
+
+          core_patches=$(git -C homebrew-core grep -h "Homebrew/formula-patches" | awk -F/ '{print $NF}' | tr -d '"')
+          all_patches=$(git ls-files | grep -v "LICENSE" | grep -v "README.md" | grep -v ".github/workflows/triage-issues.yml" | grep -v ".github/workflows/check-reference.yml")
```
Comment (Bo98) [inline .github/workflows/check-reference.yml:29]:
```suggestion
          all_patches=$(git ls-files -- '*/*' ':^.*/*')
```

This should ignore top-level files and any in a `.`-prefixed directory e.g. `.github`

File: .github/workflows/check-reference.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        uses: actions/checkout@v4
+        with:
+          repository: 'homebrew/homebrew-core'
+          path: homebrew-core
+
+      - name: Detect references
+        run: |
+          #!/bin/bash
+
+          core_patches=$(git -C homebrew-core grep -h "Homebrew/formula-patches" | awk -F/ '{print $NF}' | tr -d '"')
```
Comment (Bo98) [inline .github/workflows/check-reference.yml:28]:
```suggestion
          core_patches=$(git -C homebrew-core grep -h "Homebrew/formula-patches" | awk -F/ '{print $(NF-1)"/"$NF}' | tr -d '"')
```

You want the subdirectory too (though this strongly assumes no further nesting).

File: .github/workflows/check-reference.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+          all_patches=$(git ls-files | grep -v "LICENSE" | grep -v "README.md" | grep -v ".github/workflows/triage-issues.yml" | grep -v ".github/workflows/check-reference.yml")
+          status=0
+
+          for patch in $all_patches; do
+            if ! grep -q "$core_patches" <<< "$patch"; then
```
Comment (Bo98) [inline .github/workflows/check-reference.yml:33]:
```suggestion
            if ! grep -qx "$core_patches" <<< "$patch"; then
```

We don't want partial matches here.

**Review by Bo98 — COMMENTED** [review]
Looks good but would rather automated removal rather than yet another manual task for people to check.

**Comment by issyl0** [comment]
```
remote: Permission to Homebrew/formula-patches.git denied to github-actions[bot].
fatal: unable to access 'https://github.com/Homebrew/formula-patches/': The requested URL returned error: 403
```

**Comment by Bo98** [comment]
```yaml
permissions:
  contents: write
```

File: .github/workflows/check-reference.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+          core_patches=$(git -C homebrew-core grep -h "Homebrew/formula-patches" | awk -F/ '{print $(NF-1)"/"$NF}' | tr -d '"')
+          all_patches=$(git ls-files -- '*/*' ':^.*/*')
+
+          git checkout -b remove-unused-patches
+
+          for patch in $all_patches; do
+            if ! grep -qx "$core_patches" <<< "$patch"; then
+              echo "Unused patch: $patch; deleting."
+              git rm "$patch"
+              git commit -m "Remove unused patch: $patch"
```
Comment (Bo98) [inline .github/workflows/check-reference.yml:42]: Better matches existing commit message convention:

```suggestion
              git commit -m "${patch%%/*}: remove unused ${patch#*/}"
```

File: .github/workflows/check-reference.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+      - name: Configure Git user
+        uses: Homebrew/actions/git-user-config@master
+        with:
+          username: BrewTestBot
+
+      - name: Detect references
+        run: |
+          #!/bin/bash
+
```
Comment (Bo98) [inline .github/workflows/check-reference.yml:32]:
```suggestion
```

Don't need a shebang here

File: .github/workflows/check-reference.yml [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+on:
+  push:
+    branches-ignore:
+      - dependabot/**
+  schedule:
+    # Once every day at midnight UTC
+    - cron: "0 0 * * *"
+
+jobs:
+  stale:
```
Comment (Bo98) [inline .github/workflows/check-reference.yml:12]: `stale` seems like a copy-paste error

File: .github/workflows/check-reference.yml (hunk 6 above)
Comment (issyl0) [reply .github/workflows/check-reference.yml]: Yeah this was leftover from when I wrote this as a separate script to test it.

**Comment by SMillerDev** [comment]
https://github.com/Homebrew/formula-patches/tree/remove-unused-patches exists now. But I guess it also needs to create a PR?

**Comment by issyl0** [comment]
Or push straight to `master`? I did the branch to test it. But we could do a PR, but that also requires human intervention to merge.

File: .github/workflows/check-reference.yml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+      - name: Configure Git user
+        uses: Homebrew/actions/git-user-config@master
+        with:
+          username: BrewTestBot
+
+      - name: Detect references
+        run: |
+          core_patches=$(git -C homebrew-core grep -h "Homebrew/formula-patches" | awk -F/ '{print $(NF-1)"/"$NF}' | tr -d '"')
+          all_patches=$(git ls-files -- '*/*' ':^.*/*')
```
Comment (MikeMcQuaid) [inline .github/workflows/check-reference.yml:35]: Unclear what these are doing. A comment would be great, maybe even splitting into a dedicated script so it's easy to see the output.

Also probably want a `set -euo pipefail` so it'll actually fail if these commands do.

File: .github/workflows/check-reference.yml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        with:
+          repository: 'homebrew/homebrew-core'
+          path: homebrew-core
+
+      - name: Configure Git user
+        uses: Homebrew/actions/git-user-config@master
+        with:
+          username: BrewTestBot
+
+      - name: Detect references
```
Comment (SMillerDev) [inline .github/workflows/check-reference.yml:32]:
```suggestion
      - name: Detect references
        shell: bash
```

**Comment by Bo98** [comment]
Should be able to set up a signing step like we do in all the other repos

**Comment by SMillerDev** [comment]
I'm not sure what to do with this. @Bo98 do you want to rewrite this in ruby and call it with `brew ruby`?

**Comment by MikeMcQuaid** [comment]
@SMillerDev What's the blocker here, not sure I understand?

**Comment by SMillerDev** [comment]
The generated PR has some comments from Bo, but I'm not sure what to do with them or _if_ we should do something with them.

**Comment by MikeMcQuaid** [comment]
> The generated PR has some comments from Bo, but I'm not sure what to do with them or _if_ we should do something with them.

@SMillerDev Ok, I saw now, thanks.

Yeh, I think this needs to eliminate the false positives. Fine to not delete things for a V1, not fine to attempt to delete them when still needed.

Could always just add more conditionals to your Bash for now to exclude formulae with any head/OS/arch-specific patches or logic.

**Pull request closed without merging**